src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,721 @@
+        buf = std::move(passed_buf);
+    }
+
+    ostream(const ostream &) = delete;
+
+    void operator=(const ostream &) = delete;
+
+    ostream(const ostream &&) = delete;
+
+    void operator=(const ostream &&) = delete;
+
+    public:
+    // throw_on_states() used for throw an exception on requested "bit"
+    // flag can be: diostates::buffer_failbit and diostates::posixio_errno
+    // Multiple bits can be passed with "|" bitwise operator also.
+    void throw_on_states(const int bit, const bool do_throw) noexcept;
+
+    // Clear all situations which marked to throw exceptions
+    void clear_throws() noexcept;
+
+    // Set file descriptor
+    // NOTE: Setting new fd will REPLACE current fd in the stream and It doesn't
+    // flush data to current fd before replacing it. Also It may result in leaked
+    // file descriptor (because fstream use stream's fd to know what fd should
+    // be closed).
+    void set_fd(const int fd) noexcept;
+
+    // Get current file descriptor
+    int get_fd() noexcept;
+
+    // Returns current bits/vars in stream based on diostates enum
+    int current_state() noexcept;
+
+    // Returns 1 when there is no set bits/vars otherwise 0
+    bool good() noexcept;
+
+    // Returns true or false if buffer_failbit is set or not
+    bool buffer_failure() noexcept;
+
+    // Returns 0 or POSIX errno based on posixio_errno value
+    int posixio_failure() noexcept;
+
+    // Resets failure bits/vars to false (or 0 about posixio_errno)
+    void clear() noexcept;
+
+    // Flush buffer of stream
+    // Non _nothrow variant may throw:
+    // dio::iostream_error on Buffer failure (buffer did not allocated at construction time)
+    // dio::posixio_error on POSIX I/O failure
+    ssize_t flush_nothrow() noexcept;
+    ssize_t flush();
+
+    // write() functions, Used for write many types
+    // Non _nothrow variants may throw:
+    // dio::iostream_error on Buffer failure (buffer did not allocated at construction time)
+    // dio::posixio_error on POSIX I/O failure
+    ssize_t write_nothrow(const char *msg) noexcept;
+    ssize_t write(const char *msg);
+
+    ssize_t write_nothrow(const char msg) noexcept;
+    ssize_t write(const char msg);
+
+    ssize_t write_nothrow(const std::string &msg) noexcept;
+    ssize_t write(const std::string &msg);
+
+    ssize_t write_nothrow(const int num) noexcept;
+    ssize_t write(const int num);
+
+    ssize_t write_nothrow(const unsigned num) noexcept;
+    ssize_t write(const unsigned num);
+
+    ssize_t write_nothrow(const long num) noexcept;
+    ssize_t write(const long num);
+
+    ssize_t write_nothrow(const unsigned long num) noexcept;
+    ssize_t write(const unsigned long num);
+
+    ssize_t write_nothrow(const endline &) noexcept;
+    ssize_t write(const endline &);
+
+    ssize_t write_nothrow(const flushbuf &) noexcept;
+    ssize_t write(const flushbuf &);
+
+    template <typename T> inline ssize_t write_nothrow(T t) noexcept
```

Comment:
Hmm, Looks like for some reason the another overload try to look up a function which accepts a `char*` not `const char*`, The easiest fix for this is having overload for `char*` but It doesn't look right.

Reply:
It's not, you shouldn't need that. It should work fine if you delete the overload for `char *`.

Reply:
It's not fine with my example (because `strerror()` returns `char*` not `const char*`, WTH):
```c++
    if (!file.open(argv[1])) {
        print_err("Cannot open file for reading: ", errno, ": ", strerror(errno), ", Exiting...", endl);
        return -1;
    }
```
compiler output:
```
In file included from dio_test.cc:1:
./src/includes/dinit-iostream.h:481:24: error: no matching member function for call to 'write_nothrow'
  481 |         ssize_t ret2 = write_nothrow(b...);
      |                        ^~~~~~~~~~~~~
./src/includes/dinit-iostream.h:479:24: note: in instantiation of function template specialization 'dio::fileout::write_nothrow<char *>' requested here
  479 |         ssize_t ret1 = write_nothrow(a);
      |                        ^
./src/includes/dinit-iostream.h:481:24: note: in instantiation of function template specialization 'dio::fileout::write_nothrow<char *, char[13], dio::endline>' requested here
  481 |         ssize_t ret2 = write_nothrow(b...);
      |                        ^
./src/includes/dinit-iostream.h:481:24: note: in instantiation of function template specialization 'dio::fileout::write_nothrow<char[3], char *, char[13], dio::endline>' requested here
./src/includes/dinit-iostream.h:481:24: note: in instantiation of function template specialization 'dio::fileout::write_nothrow<int, char[3], char *, char[13], dio::endline>' requested here
./src/includes/dinit-iostream.h:509:17: note: in instantiation of function template specialization 'dio::fileout::write_nothrow<char[31], int, char[3], char *, char[13], dio::endline>' requested here
  509 |     return cerr.write_nothrow(args...);
      |                 ^
dio_test.cc:23:9: note: in instantiation of function template specialization 'dio::print_err<char[31], int, char[3], char *, char[13], dio::endline>' requested here
   23 |         print_err("Cannot open file for reading: ", errno, ": ", strerror(errno), ", Exiting...", endl);
      |         ^
./src/includes/dinit-iostream.h:472:13: note: candidate function not viable: requires 1 argument, but 0 were provided
  472 |     ssize_t write_nothrow(const flushbuf &) noexcept;
      |             ^             ~~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:469:13: note: candidate function not viable: requires 1 argument, but 0 were provided
  469 |     ssize_t write_nothrow(const endline &) noexcept;
      |             ^             ~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:466:13: note: candidate function not viable: requires single argument 'num', but no arguments were provided
  466 |     ssize_t write_nothrow(const unsigned long num) noexcept;
      |             ^             ~~~~~~~~~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:463:13: note: candidate function not viable: requires single argument 'num', but no arguments were provided
  463 |     ssize_t write_nothrow(const long num) noexcept;
      |             ^             ~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:460:13: note: candidate function not viable: requires single argument 'num', but no arguments were provided
  460 |     ssize_t write_nothrow(const unsigned num) noexcept;
      |             ^             ~~~~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:457:13: note: candidate function not viable: requires single argument 'num', but no arguments were provided
  457 |     ssize_t write_nothrow(const int num) noexcept;
      |             ^             ~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:454:13: note: candidate function not viable: requires single argument 'msg', but no arguments were provided
  454 |     ssize_t write_nothrow(const std::string &msg) noexcept;
      |             ^             ~~~~~~~~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:451:13: note: candidate function not viable: requires single argument 'msg', but no arguments were provided
  451 |     ssize_t write_nothrow(const char msg) noexcept;
      |             ^             ~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:448:13: note: candidate function not viable: requires single argument 'msg', but no arguments were provided
  448 |     ssize_t write_nothrow(const char *msg) noexcept;
      |             ^             ~~~~~~~~~~~~~~~
./src/includes/dinit-iostream.h:475:57: note: candidate function template not viable: requires at least argument 'a', but no arguments were provided
  475 |     template <typename A, typename ...B> inline ssize_t write_nothrow(const A &a, const B & ...b)
      |                                                         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
I tried to mark flags as `const` in template function but the result is still the same. `dinit-log` doesn't have any problem with this but I can't figure that how did you fixed this.

Reply:
The issue is that your templated overload accepts a single argument. So when it looks for a `char *` overload it calls itself, not the `const char *` overload.

To fix it you can write it like this:

    template <typename A, typename B, typename ...C> inline ssize_t write_nothrow(const A &a, const B &b, const C & ...c)
            noexcept
    {
        if (!good()) return -1;
        ssize_t ret1 = write_nothrow(a);
        if (ret1 < 0) return -1;
        ssize_t ret2 = write_nothrow(b, c...);
        return (ret2 > 0) ? ret1 + ret2 : ret1;
    }

That way it only accepts 2 or more arguments, and single argument calls always resolve to a non-template overload.

This isn't a problem for `dinit_log::log_parts` because the two overloads to choose from are always templated:
```
    template <typename A> static inline void log_parts(const A &a) noexcept
```
and
```
    template <typename A, typename ...B> static inline void log_parts(const A &a, const B & ...b) noexcept
```

Because they are both templates, it will choose the one without the variable-arity parameter (`B & ...b`).